lib/utils/aws/signing.go

```diff
@@ -0,0 +1,194 @@
+	}
+	return nil
+}
+
+// SigningCtx contains AWS SigV4 signing context parameters.
+type SigningCtx struct {
+	SigningName   string
+	SigningRegion string
+	Expiry        time.Time
+	SessionName   string
+	AWSRoleArn    string
+	AWSExternalID string
+}
+
+// Check checks signing context parameters.
+func (sc *SigningCtx) Check(clock clockwork.Clock) error {
+	switch {
+	case sc.SigningName == "":
+		return trace.BadParameter("missing AWS signing name")
+	case sc.SigningRegion == "":
+		return trace.BadParameter("missing AWS signing region")
+	case sc.SessionName == "":
+		return trace.BadParameter("missing AWS session name")
+	case sc.AWSRoleArn == "":
+		return trace.BadParameter("missing AWS Role ARN")
+	case sc.Expiry.Before(clock.Now()):
+		return trace.BadParameter("AWS SigV4 expiry has already expired")
+	default:
+		return nil
+	}
+}
+
+// SignRequest creates a new HTTP request and rewrites the header from the original request and returns a new
+// HTTP request signed by STS AWS API.
+// Signing steps:
+// 1) Decode Authorization Header. Authorization Header example:
+//
+//		Authorization: AWS4-HMAC-SHA256
+//		Credential=AKIAIOSFODNN7EXAMPLE/20130524/us-east-1/s3/aws4_request,
+//		SignedHeaders=host;range;x-amz-date,
+//		Signature=fe5f80f77d5fa3beca038a248ff027d0445342fe2855ddc963176630326f1024
+//
+//	 2. Extract credential section from credential Authorization Header.
+//	 3. Extract aws-region and aws-service from the credential section.
+//	 4. Build AWS API endpoint based on extracted aws-region and aws-service fields.
+//	    Not that for endpoint resolving the https://github.com/aws/aws-sdk-go/aws/endpoints/endpoints.go
+//	    package is used and when Amazon releases a new API the dependency update is needed.
+//	 5. Sign HTTP request.
+func (s *SigningService) SignRequest(req *http.Request, signCtx *SigningCtx) (*http.Request, error) {
+	if signCtx == nil {
+		return nil, trace.BadParameter("missing signing context")
+	}
+	if err := signCtx.Check(s.Clock); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	payload, err := GetAndReplaceReqBody(req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	reqCopy, err := http.NewRequest(req.Method, req.URL.String(), bytes.NewReader(payload))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	reqCopy.Header = req.Header.Clone()
+
+	unsignedHeaders := removeUnsignedHeaders(reqCopy)
+	credentials := s.GetSigningCredentials(s.Session, signCtx.Expiry, signCtx.SessionName, signCtx.AWSRoleArn, signCtx.AWSExternalID)
+	signer := NewSigner(credentials, signCtx.SigningName)
+	_, err = signer.Sign(reqCopy, bytes.NewReader(payload), signCtx.SigningName, signCtx.SigningRegion, s.Clock.Now())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// copy removed headers back to the request after signing it, but don't copy the old Authorization header.
+	copyHeaders(reqCopy, req, utils.RemoveFromSlice(unsignedHeaders, "Authorization"))
```

Comment:
+1

Reply:
I think that suggestion won't work, because we do not want all of the original request headers when we sign. If we sign with those headers, then they become signed headers as well.

Maybe I've misunderstood the suggestion though?

Reply:
Yes, that true, but why we need to add a extra headers to reqCopy form the origin req after signing reqCopy ? 
The previous version of this code the reqCopy  was jus returned:  
https://github.com/gravitational/teleport/pull/19387/files#diff-e247bd4fbfee0e25ceae30d0afae77f5f3e68f661470d5ead8519042c333ca7eL204

Reply:
In this PR, the signing service is decoupled from the app service AWS handler. Previously, we were removing app service `ReservedHeaders` and then not adding them back after signing. That is still the case, except now app service headers are removed by the app service AWS handler before giving the request to the signing service for signature. So that behavior is actually the same before/after: we remove app service headers before signing and do not copy them back.

What I've done here in the signing service is I've avoided any signature validation problems caused by signing an AWS request with headers that should not be signed, such as hop-by-hop headers (which may change before reaching AWS for signature validation). It's just a robust way of signing forwarded AWS requests. Previously, the oxy forwarder would remove hop-by-hop headers for us before we got the request to sign, but that is no longer the case since we are not signing inside `RoundTrip`.

I tested the difference in wire format sent before/after this PR just now, and there is a difference in the headers we send, which is unrelated to the signing service changes here. Before this PR, we had AWS handling logic inside `RoundTrip`, which was being called by `oxy/forward/Forwarder.ServeHTTP`. The oxy forwarder rewrites headers before calling `RoundTrip` though, using this default rewriter in `oxy/forward/rewrite.go`:
```golang
// Rewriter is responsible for removing hop-by-hop headers and setting forwarding headers
type HeaderRewriter struct {
	TrustForwardHeader bool
	Hostname           string
}
```
This means that before this PR, those forwarding headers were already injected into the request by oxy and were then removed by the AWS signing roundtripper. I think that is a mistake - we recently changed `ReservedHeaders` to include the headers injected by oxy, thus we started to remove those oxy forwarding headers. This PR fixes that - since this PR moves the AWS handling logic out of `RoundTrip`, those forwarding headers are injected by oxy after we sign the request, and so we are now sending the forwarding headers again.

Here's an example AWS request in wire format before and after these changes so you can see which headers are being sent now vs before.

Before this PR:
```
POST / HTTP/1.1
Host: localhost:8001
Accept-Encoding: identity
Authorization: AWS4-HMAC-SHA256 Credential=*snip*, SignedHeaders=accept-encoding;content-type;host;x-amz-date;x-amz-security-token;x-amz-target, Signature=*snip*
Content-Length: 2
Content-Type: application/x-amz-json-1.0
User-Agent: aws-cli/2.9.7 Python/3.11.0 Darwin/22.2.0 source/arm64 prompt/off command/dynamodb.list-tables
X-Amz-Date: 20221219T190825Z
X-Amz-Security-Token: *snip*
X-Amz-Target: DynamoDB_20120810.ListTables

{}
```

This PR:
```
POST / HTTP/1.1
Host: localhost:8001
Transfer-Encoding: chunked
Accept-Encoding: identity
Authorization: AWS4-HMAC-SHA256 Credential=*snip*, SignedHeaders=content-type;host;x-amz-date;x-amz-security-token;x-amz-target, Signature=*snip*
Content-Type: application/x-amz-json-1.0
User-Agent: aws-cli/2.9.7 Python/3.11.0 Darwin/22.2.0 source/arm64 prompt/off command/dynamodb.list-tables
X-Amz-Date: 20221219T185744Z
X-Amz-Security-Token: *snip*
X-Amz-Target: DynamoDB_20120810.ListTables
X-Forwarded-Host: localhost:8001
X-Forwarded-Port: 8001
X-Forwarded-Proto: http
X-Forwarded-Server: mac.attlocal.net

2
{}
0
```

A few things to note here about the request sent using this PR's changes:
1. `X-Forwarded-*` headers injected by oxy are now present in the request, which I believe is correct behavior because a reverse proxy should be setting those headers.
2. Content-Length is not set, so go automatically set the `Transfer-Encoding: chunked` header for us. The original request was not chunked, so this is a mistake I made. That value was actually coming from `http.Request.ContentLength`, not `http.Request.Header.Get("Content-Length")`. In fact the problem is the way I copy the request by making a new request and passing the request body directly. I will fix this by cloning the request in the AWS handler instead.
3. The signed headers no longer includes `accept-encoding`, which it shouldn't since this was not a signed header originally.

tl;dr: We remove unsigned headers and copy them back after signing so that we can sign the request without unintentionally signing other headers. I need to fix the request copying in `rewriteRequest` to use `req.Clone` instead of making a new request so that we preserve the content length in the request.

Reply:
Correction: we were not sending `x-forwarded-*` headers before the changes to `ReservedHeaders` either... I assumed we only removed those headers for signing purposes. Is there any reason we shouldn't allow the oxy forwarder to set those headers?

Reply:
IFIK no, these headers should be removed only for signing purpose.

Reply:
I agree with @smallinsky
We should include them after we sign the request

Reply:
Latest changes continue to send the oxy-injected forwarding headers and fixed the transfer encoding issue, so the request is not chunked now.